altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+        // If current is null, that means no free block was found that's large enough.
+        if current.is_null() {
+            return current as *mut u8;
+        }
+
+        let alloc_start = alignment::align_up(current as usize, using_align);
+        let align_diff = alloc_start - current as usize;
+        let current_size = unsafe { (*current).block_size };
+
+        // Block is correct size and alignment.
+        if current_size == using_size && align_diff == 0 {
+            // If at head, there is no previous to adjust
+            if self.head == current {
+                self.head = unsafe { (*self.head).next_block };
+            }
+            else {
+                unsafe { (*previous).next_block = (*current).next_block; }
+            }
+        }
+        // Current block is larger than required and has the correct alignment
+        else if current_size > using_size && align_diff == 0 {
+            unsafe { (*current).block_size -= using_size; }
+            if self.head == current {
+                self.head = self.shift_block_forward(current, using_size);
+            }
+            else {
+                unsafe {
+                    (*previous).next_block = self.shift_block_forward(current, using_size);
+                }
+            }
+        }
+        // Current block is larger than required but has exactly the right size to
+        // accomodate the alignment and size requirements
+        else if current_size == using_size + align_diff {
+            unsafe { (*current).block_size = align_diff; }
+        }
+        // Current block is larger than required and because of alignment, the allocation
+        // divides it in two.
+        else {
+            unsafe {
+                (*current).block_size = current_size - using_size - align_diff;
+                let upper_block = self.shift_block_forward(current, using_size + align_diff);
+                ptr::write(current as *mut BlockHeader, BlockHeader::new(align_diff));
+                (*current).next_block = upper_block;
+            }
+        }
+        alloc_start as *mut u8
+    }
+
+    /*
+    TODO: We can still implement merging to deal with fragmentation
+    - Nothing adjacent: Make new block, connect to closest blocks
+    - Adjacent at tail: Merge with tail block, move block, switch leading ptr
+    - Adjacent at lead: Merge with lead, no additional changes
+    - Adjacent at both: Merge two with lead (add sizes, switch lead ptr)
+    */
+
+    // Deallocates memory, placing it back in the free list for later use
+    // Adds a free block to the list based on alloc_ptr so that the list remains
+    // sorted based on memory position.
+    pub fn deallocate(&mut self, alloc_ptr: *mut u8, size: usize, _align: usize) {
+        // We can immediately add the block at the deallocated position
+        let alloc_block_ptr = alloc_ptr as *mut BlockHeader;
+        let used_memory = alignment::align_up(size, self.block_hdr_size);
```

Comment:
We're not getting the max of the two values here. This just tells us how much memory the allocation took because we use `let using_size = alignment::align_up(request_size, self.block_hdr_size);` in `allocate`.